src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -11715,6 +11737,241 @@ void Test1(object x1, object? y1)
+                //             c6._cField.GetC().ToString(); // warn 12
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c6._cField.GetC()").WithLocation(79, 13)
+            );
+        }
+
+        [Fact]
+        public void ConditionalBranching_15()
+        {
+            var compilation = CreateCompilation(@"
+class C
+{
+    void Test(C? c1)
+    {
+        if (c1?[0] != null)
+        {
+            c1.ToString();
+            c1[0].ToString(); // warn 1
+        }
+        else
+        {
+            c1.ToString(); // warn 2
+        }
+    }
+
+    object? this[int i] { get => null; }
+}
+", parseOptions: TestOptions.Regular8);
+
+            compilation.VerifyDiagnostics(
+                // (9,13): warning CS8602: Possible dereference of a null reference.
+                //             c1[0].ToString(); // warn 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c1[0]").WithLocation(9, 13),
+                // (13,13): warning CS8602: Possible dereference of a null reference.
+                //             c1.ToString(); // warn 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c1").WithLocation(13, 13)
+            );
+        }
+
+        [Fact]
+        public void ConditionalBranching_16()
+        {
+            var compilation = CreateCompilation(@"
+class C
+{
+    void Test<T>(T t)
+    {
+        if (t?.ToString() != null)
+        {
+            t.ToString();
+        }
+        else
+        {
+            t.ToString(); // warn
+        }
+    }
+}", parseOptions: TestOptions.Regular8);
+
+            // PROTOTYPE(NullableReferenceTypes): We are currently warning on all unconstrained type parameter invocations
+            // https://github.com/dotnet/roslyn/issues/28792
+            compilation.VerifyDiagnostics(
+                // (6,13): warning CS8602: Possible dereference of a null reference.
+                //         if (t?.ToString() != null)
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "t").WithLocation(6, 13),
+                // (8,13): warning CS8602: Possible dereference of a null reference.
+                //             t.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "t").WithLocation(8, 13),
+                // (12,13): warning CS8602: Possible dereference of a null reference.
+                //             t.ToString(); // warn
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "t").WithLocation(12, 13)
+                );
+        }
+
+        [Fact]
+        public void ConditionalBranching_17()
+        {
+            var compilation = CreateCompilation(@"
+class C
+{
+    object? Prop { get; }
+    object? GetObj(bool val) => null;
+    void Test(C? c1, C? c2)
+    {
+        if (c1?.GetObj(c2?.Prop != null) != null)
+        {
+            c2.Prop.ToString(); // warn 1 2
+        }
+    }
+}", parseOptions: TestOptions.Regular8);
```

Comment:
Is the following case interesting?
```
static void M(object? x, object? y)
{
    if ((x = y)?.GetHashCode() != null)
        x.ToString();
}
```

Reply:
It is, but because `GetHashCode()` returns a nullable int, and not a reference type. Should be a simple fix.

---
In reply to: [204560523](https://github.com/dotnet/roslyn/pull/28722#discussion_r204560523) [](ancestors = 204560523)